src/daemon/rpmostreed-os.cxx

```diff
@@ -932,6 +934,127 @@ os_handle_finalize_deployment (RPMOSTreeOS *interface, GDBusMethodInvocation *in
   return TRUE;
 }
 
+static DnfContext *
+os_create_bare_dnf_context (GCancellable *cancellable, GError **error)
```

Comment:
I did use it at the first, but it looked like there is too much about it. Thus I chose this "bare" thing, which is currently used in the gnome-software. I think there was something non-working with it with respect of the sack, but I do not recall what it was. It definitely failed to initialize the sack properly for me when I used some other rpm-ostree function (it used different sack flags or something).

Reply:
Sorry I know this means a bunch of work, but I'd like to dig into that a bit more. Can you try to recreate the error you were hitting? It'd be nice if we didn't have multiple code paths for creating a `DnfContext`. (We have a hack in some places that initialize one just to get the base arch because there isn't a better API, but otherwise the only place we create a `DnfContext` is in the core.)

Reply:
Using `rpmostree_context_prepare` results in:
```
Call failed: Updating rpm-md repo 'fedora-cisco-openh264': Failed to download gpg key for repo 'fedora-cisco-openh264':
Curl error (37): Couldn't read a file:// file for file:///etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-rpmostree-unset-releasever-x86_64
[Couldn't open file /etc/pki/rpm-gpg/RPM-GPG-KEY-fedora-rpmostree-unset-releasever-x86_64]
```
I get the same error when calling `rpmostree_context_download_metadata()` instead, with `DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB | DNF_CONTEXT_SETUP_SACK_FLAG_LOAD_UPDATEINFO`.

When I set `pkgcache_only` to `TRUE`, then the `GetPackages` returns nothing.

There is filled https://github.com/coreos/rpm-ostree/issues/4098 for the `rpmostree-unset-releasever`. I wrote there some info just now.

I will try to call the `rpmostree_context_setup` with both dirs set to valid values, even I'm not sure what they are (thee the #4098), but it seems it's the right way to do, to fix this one and also the `ListRepos` naming.

Reply:
I chose a different approach here, I reused the DnfContext created for the `ListRepos`. See the new commits. It addresses this thread concerns, if I'm not mistaken.